### PR TITLE
Move blockchain.tipHash, tipHeight into blockdao 1779

### PR DESCRIPTION
**Pull request by lzxm160**
work on #1779

### CODE REVIEW

File: blockchain/blockchain.go [hunk 1]
```diff
@@ -342,12 +338,14 @@ func (bc *blockchain) BlockFooterByHash(h hash.Hash256) (*block.Footer, error) {
 func (bc *blockchain) TipHash() hash.Hash256 {
 	bc.mu.RLock()
 	defer bc.mu.RUnlock()
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: delete

File: blockchain/blockchain.go [hunk 2]
```diff
@@ -386,7 +384,8 @@ func (bc *blockchain) context(ctx context.Context, tipInfoFlag, candidateFlag bo
 	var tip protocol.TipInfo
 	var err error
 	if candidateFlag {
-		if candidates, err = bc.candidatesByHeight(bc.tipHeight + 1); err != nil {
+		tipHeight, _ := bc.dao.GetTipHeight()
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: error handling

File: blockchain/blockchain.go [hunk 3]
```diff
@@ -415,8 +414,8 @@ func (bc *blockchain) MintNewBlock(
 	defer bc.mu.RUnlock()
 	mintNewBlockTimer := bc.timerFactory.NewTimer("MintNewBlock")
 	defer mintNewBlockTimer.End()
-
-	newblockHeight := bc.tipHeight + 1
+	tipHeight, _ := bc.dao.GetTipHeight()
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: error handling

File: blockchain/blockchain.go [hunk 4]
```diff
@@ -450,8 +449,7 @@ func (bc *blockchain) MintNewBlock(
 	ra := block.NewRunnableActionsBuilder().
 		AddActions(actions...).
 		Build()
-
-	prevBlkHash := bc.tipHash
+	prevBlkHash, _ := bc.dao.GetTipHash()
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: error handling

File: blockchain/blockchain.go [hunk 5]
```diff
@@ -580,11 +578,12 @@ func (bc *blockchain) startExistingBlockchain(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
-	if stateHeight > bc.tipHeight {
+	tipHeight, _ := bc.dao.GetTipHeight()
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: dito

File: blockchain/blockchain.go [hunk 6]
```diff
@@ -668,7 +665,8 @@ func (bc *blockchain) commitBlock(blk *block.Block) error {
 	if err != nil {
 		log.L().Panic("Error when committing states.", zap.Error(err))
 	}
-	blk.HeaderLogger(log.L()).Info("Committed a block.", log.Hex("tipHash", bc.tipHash[:]))
+	tipHash, _ := bc.dao.GetTipHash()
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: dito

File: blockchain/blockchain.go [hunk 7]
```diff
@@ -608,27 +607,29 @@ func (bc *blockchain) startExistingBlockchain(ctx context.Context) error {
...
+	tipHeight, _ := bc.dao.GetTipHeight()
+	if tipHeight == 0 {
 		return &protocol.TipInfo{
 			Height:    0,
 			Hash:      bc.config.Genesis.Hash(),
 			Timestamp: time.Unix(bc.config.Genesis.Timestamp, 0),
 		}, nil
 	}
-	header, err := bc.dao.Header(bc.tipHash)
+	tipHash, _ := bc.dao.GetTipHash()
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: dito

File: blockchain/blockdao/blockdao.go [hunk 8]
```diff
@@ -7,6 +7,7 @@
 package blockdao
```
Comment (CoderZhi) [inline blockchain/blockdao/blockdao.go]: comments offline

File: blockchain/blockchain.go [hunk 9]
```diff
@@ -608,27 +607,29 @@ func (bc *blockchain) startExistingBlockchain(ctx context.Context) error {
...
+	header, err := bc.dao.Header(tipHash)
 	if err != nil {
 		return nil, err
 	}
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: to handle this case, two ways:
1. dao.TipHeight(), and then use the tip height to read header
2. dao.TipHeader(), and then use the tip header to get tip height (header->Height()) and hash (header->HashBlock())

File: blockchain/blockchain.go [hunk 10]
```diff
@@ -386,7 +390,8 @@ func (bc *blockchain) context(ctx context.Context, tipInfoFlag, candidateFlag bo
 	var tip protocol.TipInfo
 	var err error
 	if candidateFlag {
-		if candidates, err = bc.candidatesByHeight(bc.tipHeight + 1); err != nil {
+		tipHeight := bc.TipHeight()
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: don't call public function in side of private function

File: blockchain/blockchain.go [hunk 11]
```diff
@@ -415,8 +420,8 @@ func (bc *blockchain) MintNewBlock(
 	defer bc.mu.RUnlock()
 	mintNewBlockTimer := bc.timerFactory.NewTimer("MintNewBlock")
 	defer mintNewBlockTimer.End()
-
-	newblockHeight := bc.tipHeight + 1
+	tipHeight := bc.TipHeight()
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: same here and all following places

File: blockchain/blockdao/blockdao.go [hunk 12]
```diff
@@ -235,12 +240,9 @@ func (dao *blockDAO) GetBlockByHeight(height uint64) (*block.Block, error) {
...
 }
 
-func (dao *blockDAO) GetTipHash() (hash.Hash256, error) {
-	return dao.getTipHash()
-}
-
 func (dao *blockDAO) GetTipHeight() (uint64, error) {
-	return dao.getTipHeight()
+	tipHeight := atomic.LoadUint64(&dao.tipHeight)
+	return tipHeight, nil
```
Comment (CoderZhi) [inline blockchain/blockdao/blockdao.go]: delete error

File: blockchain/blockchain.go [hunk 13]
```diff
@@ -340,14 +331,18 @@ func (bc *blockchain) BlockFooterByHash(h hash.Hash256) (*block.Footer, error) {
...
+		return hash.ZeroHash256
+	}
+	return tipHash
 }
 
 // TipHeight returns tip block's height
 func (bc *blockchain) TipHeight() uint64 {
-	return atomic.LoadUint64(&bc.tipHeight)
+	tipHei := bc.dao.GetTipHeight()
+	return tipHei
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: `return bc.dao.GetTipHeight()`

File: blockchain/blockchain.go [hunk 14]
```diff
@@ -608,27 +607,32 @@ func (bc *blockchain) startExistingBlockchain(ctx context.Context) error {
 	}
 	log.L().Info("Restarting blockchain.",
 		zap.Uint64("chainHeight",
-			bc.tipHeight),
+			tipHeight),
```
Comment (CoderZhi) [inline blockchain/blockchain.go]: in one line

File: blockchain/blockdao/blockdao.go [hunk 15]
```diff
@@ -293,6 +294,7 @@ func (dao *blockDAO) PutBlock(blk *block.Block) error {
 	if err := dao.putBlock(blk); err != nil {
```
Comment (CoderZhi) [inline blockchain/blockdao/blockdao.go]: wrap the whole function with 
```
dao.mutex.Lock()
defer dao.mutex.Unlock()
```

File: blockchain/blockchain.go (hunk 13 above)
Comment (lzxm160) [reply blockchain/blockchain.go]: got it

**Pull request merged**